File: hdf-diff/go/matching/srg_cci_tiebreak.go
package matching

import (
	"strings"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
)

// Composite score weights matching SAF CLI's Tier 2.
const (
	cciWeight   = 0.7
	titleWeight = 0.3
)

// SRGCCITiebreakStrategy handles ambiguous SRG matches using CCI Jaccard + token Jaccard.
// This is Tier 2 of the delta matching pipeline.
//
// Only activates for gtitle groups with multiple candidates (>1 old or >1 new).
// 1:1 matches belong to srgDeterministic (Tier 1).
type SRGCCITiebreakStrategy struct{}

// NewSRGCCITiebreakStrategy creates a new SRG CCI tiebreak matching strategy.
func NewSRGCCITiebreakStrategy() *SRGCCITiebreakStrategy {
	return &SRGCCITiebreakStrategy{}
}

// Name returns the strategy name.
func (s *SRGCCITiebreakStrategy) Name() string {
	return "srgCciTiebreak"
}

// extractCCISet extracts CCI identifiers from Tags["cci"] as a set.
// Handles both []any (from JSON unmarshal) and []string (from typed converters).
func extractCCISet(req hdf.EvaluatedRequirement) map[string]bool {
	result := make(map[string]bool)
	if req.Tags == nil {
		return result
	}
	cciVal, ok := req.Tags["cci"]
	if !ok {
		return result
	}
	switch v := cciVal.(type) {
	case []any:
		for _, c := range v {
			if str, ok := c.(string); ok {
				result[str] = true
			}
		}
	case []string:
		for _, c := range v {
			result[c] = true
		}
	}
	return result
}

// cciJaccardSimilarity computes Jaccard similarity between two CCI sets.
func cciJaccardSimilarity(a, b map[string]bool) float64 {
	if len(a) == 0 && len(b) == 0 {
		return 0
	}
	intersectionSize := 0
	for k := range a {
		if b[k] {
			intersectionSize++
		}
	}
	unionSize := len(a) + len(b) - intersectionSize
	if unionSize == 0 {
		return 0
	}
	return float64(intersectionSize) / float64(unionSize)
}

// simpleTokenJaccard computes a simple whitespace-split token Jaccard similarity.
func simpleTokenJaccard(a, b string) float64 {
	ta := simpleTokenize(a)
	tb := simpleTokenize(b)
	if len(ta) == 0 || len(tb) == 0 {
		return 0
	}
	intersectionSize := 0
	for k := range ta {
		if tb[k] {
			intersectionSize++
		}
	}
	unionSize := len(ta) + len(tb) - intersectionSize
	if unionSize == 0 {
		return 0
	}
	return float64(intersectionSize) / float64(unionSize)
}

// simpleTokenize splits on whitespace, lowercases, and returns unique tokens.
func simpleTokenize(s string) map[string]bool {
	result := make(map[string]bool)
	for _, t := range strings.Fields(strings.ToLower(s)) {
		if len(t) > 0 {
			result[t] = true
		}
	}
	return result
}

// safeTitle returns the title or empty string.
func safeTitle(req hdf.EvaluatedRequirement) string {
	if req.Title == nil {
		return ""
	}
	return *req.Title
}

// Match implements the Strategy interface.
func (s *SRGCCITiebreakStrategy) Match(oldReqs, newReqs []hdf.EvaluatedRequirement) MatchResult {
	result := MatchResult{}

	// Build gtitle → indices
	oldGtitleMap := make(map[string][]int)
	for i, req := range oldReqs {
		g := extractGtitle(req)
		if g != "" {
			oldGtitleMap[g] = append(oldGtitleMap[g], i)
		}
	}

	newGtitleMap := make(map[string][]int)
	for i, req := range newReqs {
		g := extractGtitle(req)
		if g != "" {
			newGtitleMap[g] = append(newGtitleMap[g], i)
		}
	}

	matchedOldIndices := make(map[int]bool)
	matchedNewIndices := make(map[int]bool)

	for gtitle, newIdxList := range newGtitleMap {
		oldIdxList, ok := oldGtitleMap[gtitle]
		if !ok || len(oldIdxList) == 0 {
			continue
		}

		nc := len(newIdxList)
		oc := len(oldIdxList)

		// Only handle ambiguous cases (at least one side > 1)
		if nc == 1 && oc == 1 {
			continue
		}

		// Greedy matching: process each new req, find best old candidate
		for _, ni := range newIdxList {
			if matchedNewIndices[ni] {
				continue
			}

			newCcis := extractCCISet(newReqs[ni])
			newTitle := safeTitle(newReqs[ni])

			bestIdx := -1
			bestComposite := -1.0
			bestCci := 0.0
			bestIsUnclaimed := false

			for _, oi := range oldIdxList {
				oldCcis := extractCCISet(oldReqs[oi])
				oldTitle := safeTitle(oldReqs[oi])

				cci := cciJaccardSimilarity(newCcis, oldCcis)
				title := simpleTokenJaccard(newTitle, oldTitle)
				composite := cciWeight*cci + titleWeight*title
				isUnclaimed := !matchedOldIndices[oi]

				if bestIdx == -1 ||
					(isUnclaimed && !bestIsUnclaimed) ||
					(isUnclaimed == bestIsUnclaimed && composite > bestComposite) {
					bestIdx = oi
					bestComposite = composite
					bestCci = cci
					bestIsUnclaimed = isUnclaimed
				}
			}

			if bestIdx >= 0 && bestIdx < len(oldReqs) {
				relationship := "primary"
				if matchedOldIndices[bestIdx] {
					relationship = "related"
				}

				confidence := bestCci
				if confidence < 0 {
					confidence = 0
				}

				result.Matched = append(result.Matched, MatchPair{
					OldReq:       oldReqs[bestIdx],
					NewReq:       newReqs[ni],
					Strategy:     "srgCciTiebreak",
					Confidence:   confidence,
					Relationship: relationship,
				})
				matchedOldIndices[bestIdx] = true
				matchedNewIndices[ni] = true
			}
		}
	}

	// Collect unmatched
	for i, req := range oldReqs {
		if !matchedOldIndices[i] {
			result.UnmatchedOld = append(result.UnmatchedOld, req)
		}
	}
	for i, req := range newReqs {
		if !matchedNewIndices[i] {
			result.UnmatchedNew = append(result.UnmatchedNew, req)
		}
	}

	return result
}

File: hdf-diff/src/matching/srg-cci-tiebreak.ts
import type { MatchStrategy, MatchResult, MatchPair } from './types.js';

/** Composite score weights matching SAF CLI's Tier 2. */
const CCI_WEIGHT = 0.7;
const TITLE_WEIGHT = 0.3;

/**
 * Extract CCI identifiers from a requirement's tags.cci.
 */
function extractCcis(req: Record<string, unknown>): Set<string> {
  const tags = req['tags'] as Record<string, unknown> | undefined;
  if (!tags) return new Set();
  const cci = tags['cci'];
  if (!Array.isArray(cci)) return new Set();
  return new Set(cci.filter((c): c is string => typeof c === 'string'));
}

/**
 * Extract CWE identifiers from a requirement.
 *
 * Prefers the structured `req.cwe[]` field (introduced in Evaluated_Requirement
 * Wave 1) when present and non-empty. Falls back to `tags.cwe` (array or string)
 * for compatibility with HDF files emitted before the structured field existed.
 *
 * Exported so callers and tests can verify the preference order directly.
 */
export function extractCwes(req: Record<string, unknown>): Set<string> {
  const cwe = req['cwe'];
  if (Array.isArray(cwe)) {
    const filtered = cwe.filter((c): c is string => typeof c === 'string');
    if (filtered.length > 0) return new Set(filtered);
  }
  const tags = req['tags'] as Record<string, unknown> | undefined;
  if (!tags) return new Set();
  const tagCwe = tags['cwe'];
  if (Array.isArray(tagCwe)) {
    return new Set(tagCwe.filter((c): c is string => typeof c === 'string'));
  }
  if (typeof tagCwe === 'string') {
    return new Set([tagCwe]);
  }
  return new Set();
}

/**
 * Extract tags.gtitle from a requirement.
 */
function extractGtitle(req: Record<string, unknown>): string | null {
  const tags = req['tags'] as Record<string, unknown> | undefined;
  if (!tags) return null;
  const gtitle = tags['gtitle'];
  if (typeof gtitle !== 'string' || gtitle === '') return null;
  return gtitle;
}

/**
 * Extract title from a requirement.
 */
function safeTitle(req: Record<string, unknown>): string {
  const title = req['title'];
  return typeof title === 'string' ? title : '';
}

/**
 * Jaccard similarity between two sets.
 */
function cciJaccard(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 0;
  let intersectionSize = 0;
  for (const x of a) {
    if (b.has(x)) intersectionSize++;
  }
  const unionSize = a.size + b.size - intersectionSize;
  return intersectionSize / unionSize;
}

/**
 * Simple whitespace-split token Jaccard similarity.
 */
function tokenJaccard(a: string, b: string): number {
  const ta = new Set(
    a
      .toLowerCase()
      .split(/\s+/)
      .filter((t) => t.length > 0),
  );
  const tb = new Set(
    b
      .toLowerCase()
      .split(/\s+/)
      .filter((t) => t.length > 0),
  );
  if (ta.size === 0 || tb.size === 0) return 0;
  let intersectionSize = 0;
  for (const x of ta) {
    if (tb.has(x)) intersectionSize++;
  }
  const unionSize = ta.size + tb.size - intersectionSize;
  return intersectionSize / unionSize;
}

/**
 * Create an SRG CCI Tiebreak matching strategy (Tier 2 of the delta pipeline).
 *
 * Handles ambiguous SRG matches where multiple old or new requirements share
 * the same tags.gtitle. Uses a composite score of CCI Jaccard (70%) and
 * token Jaccard on titles (30%) to pick the best match.
 *
 * Only activates for gtitle groups with multiple candidates (>1 old or >1 new).
 * 1:1 matches are left for srgDeterministic (Tier 1).
 */
export function createSrgCciTiebreakStrategy(): MatchStrategy {
  return {
    name: 'srgCciTiebreak',
    match(oldReqs: Record<string, unknown>[], newReqs: Record<string, unknown>[]): MatchResult {
      // Build gtitle → indices
      const oldGtitleMap = new Map<string, number[]>();
      for (let i = 0; i < oldReqs.length; i++) {
        const g = extractGtitle(oldReqs[i]!);
        if (g) {
          const list = oldGtitleMap.get(g);
          if (list) list.push(i);
          else oldGtitleMap.set(g, [i]);
        }
      }

      const newGtitleMap = new Map<string, number[]>();
      for (let i = 0; i < newReqs.length; i++) {
        const g = extractGtitle(newReqs[i]!);
        if (g) {
          const list = newGtitleMap.get(g);
          if (list) list.push(i);
          else newGtitleMap.set(g, [i]);
        }
      }

      const matched: MatchPair[] = [];
      const matchedOldIndices = new Set<number>();
      const matchedNewIndices = new Set<number>();

      for (const [gtitle, newIdxList] of newGtitleMap) {
        const oldIdxList = oldGtitleMap.get(gtitle);
        if (!oldIdxList || oldIdxList.length === 0) continue;

        const nc = newIdxList.length;
        const oc = oldIdxList.length;

        // Only handle ambiguous cases (at least one side > 1)
        // 1:1 is handled by srgDeterministic
        if (nc === 1 && oc === 1) continue;

        // Greedy matching: process each new req, find best old candidate
        for (const ni of newIdxList) {
          if (matchedNewIndices.has(ni)) continue;

          const newCcis = extractCcis(newReqs[ni]!);
          const newTitle = safeTitle(newReqs[ni]!);
          const newCwes = extractCwes(newReqs[ni]!);

          let bestIdx = -1;
          let bestComposite = -1;
          let bestCci = 0;
          let bestCwe = -1;
          let bestIsUnclaimed = false;

          for (const oi of oldIdxList) {
            const oldCcis = extractCcis(oldReqs[oi]!);
            const oldTitle = safeTitle(oldReqs[oi]!);
            const oldCwes = extractCwes(oldReqs[oi]!);

            const cci = cciJaccard(newCcis, oldCcis);
            const title = tokenJaccard(newTitle, oldTitle);
            // CWE Jaccard is used purely as a tiebreaker — it does NOT
            // change composite scoring weights, so existing tests that
            // depend on CCI+title ordering are unaffected.
            const cwe = cciJaccard(newCwes, oldCwes);
            const composite = CCI_WEIGHT * cci + TITLE_WEIGHT * title;
            const isUnclaimed = !matchedOldIndices.has(oi);

            // Prefer unclaimed candidates; among same claim status, prefer
            // higher composite; among equal composites, prefer higher CWE
            // Jaccard (tiebreaker added Wave 2).
            if (
              bestIdx === -1 ||
              (isUnclaimed && !bestIsUnclaimed) ||
              (isUnclaimed === bestIsUnclaimed && composite > bestComposite) ||
              (isUnclaimed === bestIsUnclaimed &&
                composite === bestComposite &&
                cwe > bestCwe)
            ) {
              bestIdx = oi;
              bestComposite = composite;
              bestCci = cci;
              bestCwe = cwe;
              bestIsUnclaimed = isUnclaimed;
            }
          }

          if (bestIdx >= 0) {
            const relationship = matchedOldIndices.has(bestIdx) ? 'related' : 'primary';
            matched.push({
              oldReq: oldReqs[bestIdx]!,
              newReq: newReqs[ni]!,
              strategy: 'srgCciTiebreak',
              confidence: Math.max(bestCci, 0),
              relationship,
            });
            matchedOldIndices.add(bestIdx);
            matchedNewIndices.add(ni);
          }
        }
      }

      // Collect unmatched
      const unmatchedOld: Record<string, unknown>[] = [];
      for (let i = 0; i < oldReqs.length; i++) {
        if (!matchedOldIndices.has(i)) unmatchedOld.push(oldReqs[i]!);
      }
      const unmatchedNew: Record<string, unknown>[] = [];
      for (let i = 0; i < newReqs.length; i++) {
        if (!matchedNewIndices.has(i)) unmatchedNew.push(newReqs[i]!);
      }

      return { matched, unmatchedOld, unmatchedNew };
    },
  };
}
